suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
suppressMessages(library(tidyr))
suppressMessages(library(cowplot))
suppressMessages(library(gridExtra))

options(dplyr.summarise.inform = FALSE)
options(warn = -1)
debuggingState(on = FALSE)

file_finder = function(dir) {
  # search a given folder for .bam, .vcf, .fasta files and match to sample name
  # dir   :   directory to search for
  
  # returns  - dataframe(sample_name, gene, vcf_path, bam_path, fasta_path)
  
  samples = basename(list.dirs(dir, recursive = F))
  samples = samples[!(samples %in% c("renamed", "figures", "bySubtype"))]
  
  out_df = data.frame(samples = samples, stringsAsFactors = F)
  out_df$vcf = rep(NA, nrow(out_df))
  out_df$fasta = rep(NA, nrow(out_df))
  out_df$alleleDepth = rep(NA, nrow(out_df))
  
  for (s in samples) {
    # vcf
    tmp_vcf = list.files(path = paste0(dir, "/", s), pattern = "*.vcf$", full.names = T)
    out_df$vcf[which(out_df$samples == s)] = list(tmp_vcf)
    
    # fasta
    tmp_fasta = list.files(path = paste0(dir, "/", s), pattern = "*.fasta$", full.names = T)
    tmp_fasta = tmp_fasta[!(tmp_fasta %in% c(paste0(dir,"/", s, "/contigs.fasta")))]
    out_df$fasta[which(out_df$samples == s)] = list(tmp_fasta)
    
    # read depth file
    tmp_aa = list.files(path = paste0(dir, "/", s, "/", "tables"), pattern = "*allAlleles.txt$", full.names = T)
    out_df$alleleDepth[which(out_df$samples == s)] = list(tmp_aa)
  }

  out_df = out_df %>% 
    pivot_longer(cols = -samples, names_to = "ftype", values_to = "full_path") %>% 
    unnest_longer(full_path) %>%
    mutate(
      fname = case_when(
        ftype %in% c('vcf', 'fasta') ~ gsub("\\..+$", "", basename(full_path)),
        ftype == 'alleleDepth' ~ gsub("-allAlleles.txt", "", basename(full_path)),
        TRUE ~ basename(full_path)
        )
      ) %>%
    rename(sample_name = samples)
  out_df$full_name = paste(out_df$sample_name, out_df$fname, sep = "/")
  out_df_wide = out_df %>% pivot_wider(names_from = c(ftype), values_from = full_path) %>% drop_na(alleleDepth)
  
  return(out_df_wide)
  }

get_data_location <- function(base_location, type) {
  # searches a given directory and returns names table files generated by IRMA.
  pats <- list(
    rc = "READ_COUNTS.txt", del = "*-deletions.txt", var = "*-variants.txt",
    aa = "*-allAlleles", ins = "*-insertions.txt", cov = "*-coverage.txt",
    ps = "*-paringStats.txt",  a2m = "*coverage.a2m.txt$"
  )

  if (type %in% pats) {
    message("Error - Could not find tables to parse from pattern: " + paste0(type))
  } else {
    pattern <- pats[[type]]
  }
  f <- list.files(base_location, pattern = pattern, full.names = TRUE, recursive = TRUE)
  return(f)
}

get_minor_vars <- function(file_location, ...) {
  # reads in vcf file to get minor variants
  # return df of: pos ref alt dp af
  extra <- list(...)
  tryCatch(
    vcf <- read.delim(file_location,
      header = F, skip = 21, sep = "\t",
      col.names = c("chrom", "pos", "id", "ref", "alt", "qual", "filter", "info"),
      colClasses = "character"
    ),
    error = function(e) {
      message(paste0("Error, VCF file doesn't exist in", file_location))
      return(0)
    }
    # finally = {message("Error, VCF file doesn't exist");return(0)}
  )
  if (nrow(vcf) == 0) {
    #write("VCF file is empty \n", stdout())
    return(0)
  }

  vcf <- vcf %>%
    filter(filter == "PASS") %>%
    separate(
      col = info, into = c("dp", "af", "aq", "pub", "qub"),
      sep = ";", extra = "merge"
    )
  vcf$dp <- as.numeric(gsub(x = vcf$dp, pattern = "DP=", replacement = ""))
  vcf$af <- as.numeric(gsub(x = vcf$af, pattern = "AF=", replacement = ""))
  vcf$pos <- as.numeric(vcf$pos)

  output <- vcf[, c("pos", "ref", "alt", "dp", "af")]
  return(output)
}

get_data <- function(file_location, file_name) {
  # returns a list of dataframes read from a list of file names

  dat_out <- list()
  for (i in seq_along(file_location)) {
    tryCatch(
      expr = {
        df <- read.delim(file_location[i], sep = "\t", stringsAsFactors = FALSE)
        if (nrow(df) != 0) {
          data_avg = average_counts(df, 'aa', 5)
          dat_out[[ file_name[i] ]] = data_avg
          }
        },
      error = function(cond){
        dat_out <- append(dat_out, list(data.frame()))
      }
    )
  }
  return(dat_out)
}

average_counts <- function(df, type, by_num) {
  # type    : the aforementioned file patterns: rc, del, var, aa, ins, cov, ps
  # by_num  : the number of rows to average by
  # returns : a dataframe similar to input but averaged over by_num

  ## type = aa
  if (type == "aa") {
    df_out <- df %>%
      filter(Allele_Type == "Consensus") %>%
      group_by(group_var = trunc(2:(n() + 1) / by_num)) %>%
      summarise(coverage = mean(Count), genomic_position = min(Position))
  } else {
    return("probably an error")
  }
  return(df_out)
}

plot_flu_cov <- function(data_aa, sample_name, gene, vcf) {
  # plots coverage given a dataframe
  # df        :   dataframe  - output of average_counts function
  # sample    :   sample name
  # gene      :   gene name
  # vcf       :   vcf location
  
  minor_vars <- get_minor_vars(vcf)
  plot_name <- paste(sample_name, gene, sep = "/")
  # catch empty or missing vcf files
  if (minor_vars == 0 | minor_vars == 00) {
    annotations <- list()
  } 
  else {
    minor_loc = minor_vars$pos
    minor_dp = round(minor_vars$dp * minor_vars$af, digits = 1)
    snp = paste0(minor_vars$ref, minor_vars$pos, minor_vars$alt, " :DP", minor_dp)

    y_centre = mean(range(data_aa$coverage))
    y_max = max(data_aa$coverage) - ((y_centre / 20) * 3)
    x_minor_unit = mean(range(data_aa$genomic_position)) / 30

    annotations <- list(
      geom_vline(xintercept = minor_loc, color = "grey"),
      annotate(
        geom = "text", x = minor_loc,
        y = y_max, label = snp, color = "black", angle = 90, size = 3
      )
    )
  }

  p = ggplot(data_aa, aes(x = genomic_position, y = coverage)) +
    geom_col(fill = gcolor[[gene]]) +
    annotations +
    scale_y_continuous(expand = c(0, 0)) +
    scale_x_continuous(expand = c(0, 0)) +
    ggtitle(paste(plot_name)) +
    coord_cartesian(clip = "off")

  return(p)
}

plot_rsv_cov <- function(data_aa, plot_name, vcf) {
  # plots coverage given a dataframe.
  # data_aa      :   dataframe - output of average_counts()
  # plot_name    :   character of plot name

  # catch empty or missing vcf files
  tryCatch(expr = {
    minor_vars <- get_minor_vars(vcf)}, 
    error = function(e) {
     stop(message(paste0("Error, sample names contains illegal characters, please remove: ", plot_name)))
    }
  )

  if (minor_vars == 0 | minor_vars == 00) {
    annotations <- list()
  } else {
    minor_loc <- minor_vars$pos
    minor_dp <- round(minor_vars$dp * minor_vars$af, digits = 1)
    snp <- paste0(minor_vars$ref, minor_vars$pos, minor_vars$alt, " - DP:", minor_dp)

    y_centre <- mean(range(data_aa$coverage))
    y_max <- max(data_aa$coverage) - ((y_centre / 20) * 3)
    x_minor_unit <- mean(range(data_aa$genomic_position)) / 30

    annotations <- list(
      geom_vline(xintercept = minor_loc, color = "black", size = 0.25),
      annotate(
        geom = "text", x = minor_loc,
        y = y_max, label = snp, color = "black", angle = 90, size = 2.5))
  }

  p <- ggplot(data_aa, aes(x = genomic_position, y = coverage)) +
    # primer regions
    geom_rect(data = rsv_primer, inherit.aes = FALSE,
              aes(xmin = loc_start,
                  xmax = loc_end,
                  ymin = -Inf, ymax = Inf, 
                  fill = c("red", "green", "blue", "yellow")),
              alpha = 0.3) +
    
    geom_col() +
    geom_hline(yintercept = -50) +
    # genes displayed below
    geom_rect(data = rsv_gene_locs, inherit.aes = FALSE,
              aes(xmin = loc_start,
                  xmax = loc_end, 
                  ymin = 0, ymax = -100), 
              color = c("red", "grey", "green", "yellow", "red", "grey", "green", "yellow", "red", "grey"),
              fill = c("red", "grey", "green", "yellow", "red", "grey", "green", "yellow", "red", "grey")) +
    # gene names
    geom_text(data = rsv_gene_locs,
              aes(x = loc_start + (0.05 * loc_start), 
                  y = -50, 
                  label = gene_name), 
              size = 3) +

    annotations +
    scale_y_continuous(expand = c(0, 0)) +
    scale_x_continuous(expand = c(0, 0)) +
    ggtitle(paste(plot_name)) +
    coord_cartesian(clip = "off") + 
    theme(legend.position = "none")

  return(p)
}

plot_combine_rsv <- function(file_names) {
  # combines plots into one neat page
  # file_names  :   dataframe containing info on file names and locations and data_aa list
  
  future::plan(future::multisession, workers = 4)
  plots = furrr::future_map(.x = 1:nrow(file_names),
                            .f = function(i){
                              plot_rsv_cov(data_aa = file_names$data_aa[[i]],
                                           plot_name = file_names$sample_name[i],
                                           vcf = file_names$vcf[i])
                            }
  )

  nCol <- 4
  nRow <- 2
  plots_combined <- marrangeGrob(plots, ncol = nCol, nrow = nRow)
  return(plots_combined)
}


plot_combine_flu <- function(file_names) {

  future::plan(future::multisession, workers = 4)
  plots = furrr::future_map(.x = 1:nrow(file_names),
                            .f = function(i){
                              plot_flu_cov(data_aa = file_names$data_aa[[i]],
                                           sample_name = file_names$sample_name[i],
                                           gene = file_names$fname[i],
                                           vcf = file_names$vcf[i])
                              }
                            )

  n <- length(plots)
  nCol <- 4
  nRow <- 2
  plots_combined <- marrangeGrob(plots, ncol = nCol, nrow = nRow)
  return(plots_combined)
}