### there is an issue with the two plot functions for RSV/FLU. 
### They are not harmonised, i.e. they have different inputs. Make sure the inputs are identical for easy use.

suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
suppressMessages(library(tidyr))
suppressMessages(library(stringr))
suppressMessages(library(cowplot))
suppressMessages(library(gridExtra))

options(dplyr.summarise.inform = FALSE)
options(warn=-1)
debuggingState(on=FALSE)

get_data_location = function(base_location, type){
  # searches a given directory and returns names table files generated by IRMA.
  pats = list(rc = 'READ_COUNTS.txt', del = '*-deletions.txt', var = '*-variants.txt', 
              aa = '*-allAlleles', ins = '*-insertions.txt', cov = '*-coverage.txt', 
              ps = '*-paringStats.txt')
  
  if (type %in% pats) {
    message("Error - Could not find tables to parse from pattern: " + paste0(type))} else{
      pattern = pats[[type]]
    }
  f = list.files(base_location, pattern = pattern, full.names = TRUE, recursive = TRUE)
  return(f)
}

get_minor_vars = function(file_location, ...) {
  # reads in vcf file to get minor variants
  # return df of: pos ref alt dp af
  extra = list(...)

  tryCatch(
    vcf <- read.delim(file_location, header = F, skip =  21, sep = "\t", 
                     col.names = c("chrom", "pos", "id", "ref", "alt", "qual", "filter", "info"),
                     colClasses = "character")
    #error = function(e){message("Error, VCF file doesn't exist");return(0)},
    #finally = {message("Error, VCF file doesn't exist");return(0)}
    )
  if ( nrow(vcf) == 0) {cat("VCF file is empty \n"); return(0) }
  
  vcf = vcf %>%
    filter(filter == "PASS") %>%
    separate(col = info, into = c("dp", "af", "aq", "pub", "qub"),
             sep = ";", extra = "merge")
  vcf$dp = as.numeric(gsub(x = vcf$dp, pattern = "DP=", replacement = ""))
  vcf$af = as.numeric(gsub(x = vcf$af, pattern = "AF=", replacement = ""))
  vcf$pos = as.numeric(vcf$pos)
  
  output = vcf[, c("pos", "ref", "alt", "dp", "af")]
  return(output)
}

#get_minor_vars("~/Desktop/flu-data/W-WILDTYPE_S3/A_HA_H1.vcf")

get_data = function(file_location) {
  # returns a list of dataframes read from a list of file names
  
  dat_out = list()
  for (f in file_location) {

    df <- read.delim(f, sep = "\t", stringsAsFactors = FALSE)
    if(nrow(df) != 0) {
      splits = unlist(str_split(string = f, pattern = "/", simplify = F))
      fsplit = gsub(splits[8], pattern = "-allAlleles.txt", replacement = "")
      df$sampleID = paste0(splits[6], "/", fsplit)
      dat_out <- append(dat_out, list(df))
      }
  }
  return(dat_out)
}

average_counts = function(df, type, by_num) {
  # type    : the aforementioned file patterns: rc, del, var, aa, ins, cov, ps
  # by_num  : the number of rows to average by
  # returns : a dataframe similar to input but averaged over by_num 
  
  ## type = aa
  if (type == 'aa'){
    df_out = df %>% 
      filter(Allele_Type == "Consensus") %>% 
      group_by(group_var = trunc(2:(n()+1)/by_num)) %>%
      summarise(coverage = mean(Count), genomic_position = min(Position))
    } else {
      return("this is tmp, probably an error")
    }
  return(df_out)
}

plot_flu_cov <- function(df, sample, gene) {
  # plots coverage given a dataframe
  # df        :   dataframe  - output of average_counts function
  # plot_name :   character of plot name
  
  vcf = list.files(path = paste(base_location, sample, sep = "/"), 
                   pattern = paste0(gene, ".vcf"), full.names = T)
  minor_vars = get_minor_vars(vcf)
  plot_name = paste(sample, gene, sep = "/")
  
  #catch empty or missing vcf files
  if (minor_vars == 0 | minor_vars == 00) {
    annotations = list()
    
  } else {
    minor_loc = minor_vars$pos
    minor_dp = round(minor_vars$dp * minor_vars$af, digits = 1)
    snp = paste0(minor_vars$ref, minor_vars$pos, minor_vars$alt, ":DP", minor_dp)
    
    y_centre = mean(range(df$coverage))
    y_max = max(df$coverage) - ((y_centre/20)*3)
    x_minor_unit = mean(range(df$genomic_position))/30
    
    annotations = list(
      geom_vline(xintercept = minor_loc, color = "grey"),
      annotate(geom = "text", x = minor_loc, 
               y = y_max, label = snp, color = 'black', angle = 90, size = 2.5))
  }
  
  p = ggplot(df, aes(x = genomic_position, y = coverage)) + 
    geom_col(fill = gcolor[[gene]]) + 
    annotations +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0)) + 
    ggtitle(paste(plot_name)) + 
    coord_cartesian(clip = 'off')
  
  return(p)
}

#plot_flu_cov(df = df, sample = "D-DUAL-MUTANT_S4", gene = gene)

plot_rsv_cov <- function(df, plot_name) {
  # plots coverage given a dataframe.
  # df        :   dataframe  - output of average_counts function
  # plot_name :   character of plot name
  # minor_var :   df with minor vars
  minor_vars = get_minor_vars(base_location = base_location, sampleID = plot_name)

  #catch empty or missing vcf files
  if (minor_vars == 0 | minor_vars == 00) {
    annotations = list()
  } else{
    
    minor_loc = minor_vars$pos
    minor_dp = round(minor_vars$dp * minor_vars$af, digits = 1)
    snp = paste0(minor_vars$ref, minor_vars$pos, minor_vars$alt, ":DP", minor_dp)

    y_centre = mean(range(df$coverage))
    y_max = max(df$coverage) - ((y_centre/20)*3)
    x_minor_unit = mean(range(df$genomic_position))/30
    
    annotations = list(
      geom_vline(xintercept = minor_loc, color = "steelblue1"),
      annotate(geom = "text", x = minor_loc, 
               y = y_max, label = snp, color = 'black', angle = 90, size = 2.5))
  }

  p = ggplot(df, aes(x = genomic_position, y = coverage)) + 
    geom_col() + 
    annotations +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0)) + 
    ggtitle(paste(plot_name)) + 
    coord_cartesian(clip = 'off')
  
  return(p)
}

#plot_rsv_cov(data_aa_avg$N90000634_S11, plot_name = "N90000634_S11")

plot_combine = function(data_avg, plot_func, sample, gene) {
  # combines plots into one neat page
  # data_avg    :   averaged and combined list of data
  # plot_func   :   the plotting function, organisms specific
  # sample      :   sample name
  
  plots = list()
  for (i in seq_along(data_avg)) {
    
    p = plot_func(df = data_avg[[i]], 
                  sample = sample[[i]], 
                  gene = gene[[i]])
    
    plots = append(plots, list(p))
  }
  
  n = length(plots)
  #nCol = floor(sqrt(n))
  nCol = 4
  #nRow = ceiling(sqrt(n))
  nRow = 2
  plots_combined = marrangeGrob(plots, ncol = nCol, nrow = nRow)
  return(plots_combined)
}

# usage example is here:
#base_location = '~/Desktop/flu-data'

#locations_aa = get_data_location(base_location, type = 'aa')
#data_aa = get_data(locations_aa)
#data_aa_avg = lapply(data_aa, FUN = average_counts, type = 'aa', by_num = 5)


